elastic/apm-server
# Add support for receiving traces over OTLP/gRPC

## Motivation/summary

Add the gRPC service for receiving OTLP traces, and plumb it through to `processor/otel`. Once this lands we can follow up with:
 - adding support for metrics
 - translating OpenTelemetry semantic conventions to our model (i.e. what's in elasticexporter in opentelemetry-collector-contrib)
 - (later) adding support for logs

We use an interceptor for ensuring clients are authorised. Auth works by setting the "Authorization" header (gRPC request metadata) just like in HTTP requests.

## Checklist

~- [ ] Documentation has been updated~ (I will create a separate issue for this)

## How to test these changes

1. Run APM Server
2. Run opentelemetry-collector with the "otlp" exporter, pointed at APM Server, using the standard port (8200). It will be necessary to set `insecure: true` in the exporter config, unless APM Server is configured to use TLS.
3. Send some OpenTelemetry data to opentelemetry-collector, and ensure that it shows up in the APM app.
4. Configure APM Server with a secret token.
5. Run step 3 again, observe a failure due to invalid auth.
6. Add the secret token to the otlp exporter config, like `headers: {'Authorization': 'Bearer secret_token'}`, restart opentelemetry-collector
7. Run step 3 again, ensure the data shows up in the APM app again.

## Related issues

#4503
